Kind local-up-cluster from source implementation, for linux native de…

jayunit100 opened the pull request:
This adds a `Kind` developement environment to calico.  This is meant for people who are doing more advanced calico work and already have a linux machine w/ kubernetes set up. 

The existing Vagrant recipe can be used to bootstrap a development environment if someone doesn't already have one.  WE can phase the existing Vagrant one out eventually if nobody needs it, or refactor it to *only* bootstrap the VM, and then use kind for all other testing.

caseydavenport commented:
Deploy preview for *calico* ready!

Built with commit 090b5452a0067404e0996cd9ba58c1c3af7e9528

https://deploy-preview-2994--calico.netlify.com

caseydavenport commented:
Deploy preview for *calico* ready!

Built with commit 30d67841cbab03215fcf9bb37230ebb6407df559

https://deploy-preview-2994--calico.netlify.com

lwr20 commented on hack/development-environment/kind-local-up.sh:
```diff
@@ -0,0 +1,76 @@
+            kubectl apply -f ./calico.yaml 
+            kubectl get pods -n kube-system
+        popd
+    sleep 5 ; kubectl -n kube-system set env daemonset/calico-node FELIX_IGNORELOOSERPF=true
```
This isn't a great idea - it changes the FELIX settings in a way which wouldn't be set on a 'real' cluster.  I think 
```
kind get nodes | xargs -n1 -I {} docker exec {} sysctl -w net.ipv4.conf.all.rp_filter=0
```
is better (i.e. set RPF to `none` on all the kind nodes in a way which allows felix to set `strict` on the pod interfaces)

lwr20 replied:
According to https://github.com/kubernetes-sigs/kind/issues/891#issuecomment-541119211 - even this shouldn't be needed for too much longer

caseydavenport replied:
@lwr20 the goal of this script seems to be a development environment.

I agree this is absolutely not recommended for production, and we probably don't want it in our test infrastructure, but this should be OK in a development environment. If / when we move to using the script for critical testing / regressions, we can remove this. I'm OK leaving it in for now though.

caseydavenport replied:
(That said I'm also OK with us adjusting the node settings as you described, so long as that is reliable)

jayunit100 commented:
ok cool ! ya lets start here and then just tell me what knobs you want ill put them in over time.

jayunit100 commented:
@caseydavenport   made 2 minor updates since your review ........
- i implemented the `BUILD=true/false` option you mentioned .  agree no need to build from source every time.
- also moved the `kind` cluster into a 'calico-test` cluster name so it doesnt interfere w/ other clusters. can parameterize that later if theres a use case
